<!DOCTYPE html>
<html>
  <head>
    <title>Interactive Haplotype viewer.</title>
    <script type="text/javascript" src="protovis-r3.2.js"></script>
    <script type="text/javascript" src="seq.json"></script>
    <link type="text/css" rel="stylesheet" href="button.css"/>
    </head>
  <body>
    <h1>Interactive haplotype viewer.</h1>

    Region:
    <form name="region" action="javascript:ongo()" method="post">
      <select name="chrom">
        <option value="Chr1">Chr1</option>
        <option value="Chr2">Chr2</option>
      </select>
      <input name="start" type="number" placeholder="start" value="1">
      <input name="end" type="number" placeholder="end" value="10000">
      <input name="go" type="submit" value="Go">

    </form>

    <div id="pv_canvas">
      <canvas></canvas>
    </div>

    <script type="text/javascript+protovis">

      //function which extracts features for specified region
      var seq2feature_seq = function(chrom, start, end, seq) {
        // create object for storing features
        var feature_seq = {};
        for (var i in seq) {
          var feature = seq[i];
          var type;
          if (feature.type.match(/^SNP_.*/)) {
            type = "SNP";
          } else {type = feature.type};
          var region = {'start': start, 'end': end};
          var feat_pos = {'start': parseInt(feature.start), 'end': parseInt(feature.end)};

          if ( (feature.seqid === chrom)
            && (feat_pos.start <= region.end) && (feat_pos.end >= region.start) ){
              if (!feature_seq[type]) {
                feature_seq[type] = [];
              };
              feature_seq[type].push(feature);
            };
        };
        return feature_seq;
      };

      // function for creating strains object
      // takes a list of feature objects
      var construct_strains = function(strain_list){
        var strains = {};
        for (var s in strain_list){
          var strain = strain_list[s];
          // extract name of the strain
          var name = strain.attributes.Strain.split("=")[1];
          if (!strains[name]) {strains[name] = {seq: [], snps: []};};
          var base = strain.attributes.Change.split(":")[1];
          strains[name].seq[strain.start] = base;
          strains[name].snps.push(strain);
        }
        return strains;
      }

      // function for condtructing haplotypes
      var construct_haplotypes = function(strains){
        var haplo_hash = {};
        // first construct hash for finding haplotypes
        for (var s in strains){
          var strain = strains[s];
          var seq_id = strain.seq.toString();
          if (!haplo_hash[seq_id]){
            haplo_hash[seq_id] = [];
          }
          haplo_hash[seq_id].push(strain);
        }
        // put all haplotypes in an array
        var haplotypes = [];
        for (h in haplo_hash){
          haplotypes.push(haplo_hash[h]);
        }
        return haplotypes;
      };

      // visualization function
      var visualize = function(feature_seq, strains, haplotypes, start, end){

        // calculate number of tracks
        var n_non_snps = Object.keys(feature_seq).length -1;
        var n_haplotypes = haplotypes.length;
        var n_tracks = n_non_snps + n_haplotypes;

        // features might overlap,
        // then, it is needed to split them in separate tracks
        
        
        // view size
        var view = {
          htrack: 25,
          hbar: 10,
          track_margin: 5,
          left: 25,
          right: 25,
          top: 50,
          bottom: 25,
          width: 750
        };

        view.height = view.htrack * n_tracks + view.top + view.bottom;
        // variable for storing where last added track ends
        view.end = view.top;
        // create a panel
        var vis = new pv.Panel()
        .canvas('pv_canvas')
        .width(view.width)
        .height(view.height)
        .top(view.top)
        .bottom(view.bottom);

        // make a Scale for sequence
        var seqScale = pv.Scale.linear(start, end)
        .range(view.left, view.width - view.right);

        // make a ruller showing sequence scale
        var seqRule = vis.add(pv.Rule)
        .data(seqScale.ticks())
        .left(seqScale)
        .height(view.height - view.bottom)
        .strokeStyle("gray")
        .anchor("top").add(pv.Label)
        .textAlign("center");

        function addTrack(color, feat){
          if (feat === undefined) {
            return;
          };
          vis.add(pv.Bar)
          .data(feat)
          .height(view.hbar)
          .top(view.end + view.track_margin)
          .left(function(d) seqScale(d.start))
          .width(function(d) seqScale(d.end) - seqScale(d.start))
          .fillStyle(pv.colors(color))
          .anchor("left").add(pv.Label)
          .text(function(d) d.attributes.Name.split("=")[1]);
          
          view.end += view.htrack;
        };

        
        // make track for genes
        addTrack("lightgreen", feature_seq.gene)
        //make track for mRNA
        addTrack("lightblue", feature_seq.mRNA);

        // determine color for snp
        var snpColor = {
          'A': pv.color('red'),
          'C': pv.color('green'),
          'G': pv.color('blue'),
          'T': pv.color('orange'),
        };

        // make haplotype tracks
        function addHaplotypeTracks() {
          //reapeate for each haplotype
          for (h in haplotypes) {
            var haplotype = haplotypes[h];

            // make a baseline pvRuller
            vis.add(pv.Bar)
            .top(view.end + view.track_margin)
            .height(view.hbar)
            .left(view.left)
            .right(view.right)
            .fillStyle("lightgrey");

            vis.add(pv.Dot)
            .data(haplotype[0].snps)
            .top(view.end + view.hbar/2 + view.track_margin)
            .left(function(d) seqScale(d.start))
            .shape("triangle")
            .strokeStyle(function(d) snpColor[d.attributes.Change.split(":")[1]])
            .fillStyle(function(d) snpColor[d.attributes.Change.split(":")[1]]);;

            view.end += view.htrack;
          }; 
        };
        addHaplotypeTracks();
        
        // render image
        vis.render();
      };

      // function which is run when user submits the form
      var ongo = function() {

        // get start, end and chromosome values from the form
        var chrom = document.forms.region.chrom.value;
        var start = parseInt(document.forms.region.start.value);
        var end = parseInt(document.forms.region.end.value);

        // get feature sequence for the region
        var feature_seq = seq2feature_seq(chrom, start, end, seq);
        var snps = feature_seq.SNP;
        var strains = construct_strains(snps);
        var haplotypes = construct_haplotypes(strains);
        console.log(feature_seq);
        // start visualization
        visualize(feature_seq, strains, haplotypes, start, end);
      };

  </script>
  </body>
</html>

